Feature/multiple processes

boylea opened the pull request:
- [x] issues: fixes #4363
- [x] tests added / passed
- [ ] release document entry

This PR is to continue the work done in PR #4364, which addresses Issue #4363 

From the Tornado docs:

> Note that multiple processes are not compatible with the autoreload
> module (or the ``autoreload=True`` option to `tornado.web.Application`
> which defaults to True when ``debug=True``).
> When using multiple processes, no IOLoops can be created or
> referenced until after the call to ``TCPServer.start(n)``.

I'd like to think more about how this affects bokeh and the way it references IOLoops

boylea commented on bokeh/server/application_context.py:
```diff
@@ -18,7 +18,7 @@
     def __init__(self, application_context):
         self.application_context = application_context
-        self._callbacks = _CallbackGroup() #self.application_context.io_loop)
+        self._callbacks = _CallbackGroup(self.application_context.io_loop)
```
I changed this back to make this issue work. Thus, I'm curious, @quasiben what you were trying to do here?

boylea commented:
Rebased to fix merge conflicts (and some personal screw-ups). Ready for review

bryevdv commented on bokeh/server/tornado.py:
```diff
@@ -252,8 +248,28 @@ def __init__(self, applications, prefix, hosts,
+
         self._clients = set()
         self._executor = ProcessPoolExecutor(max_workers=4)
+        print(self._loop)
```
can probably remove this

boylea replied:
ha, I had removed and introduced it back in when I rebased. I'll take it back out

bryevdv commented:
So this looks good, I do want to try and take some time to evaluate @havocp 's main observation, which IIRC is just that the process `fork` needs to happen before most anything else (e.g. application lifecycle callbacks, etc) I think this is the case but it's worth more scrutiny. I'm on a short vacation until Tuesday but I will try to review things more closely with that concern in mind by mid-week next week. Of course, you can take a look into this as well if you have the time. Thanks for the PR!

boylea commented:
I added `num_procs` to the doc string, I don't know if the format is correct though

birdsarah commented on bokeh/server/tests/test_server.py:
```diff
@@ -4,9 +4,12 @@
 import logging
 import re
 
+from unittest.mock import patch
```
Need to use import mock to make this py2 compatible

boylea replied:
Done

havocp commented on bokeh/command/subcommands/serve.py:
```diff
@@ -124,6 +124,16 @@
+
+.. code-block:: sh
+
+    bokeh serve app_script.py --num_procs 2
```
`--num-procs` vs. `--num_procs` here

boylea replied:
Done

havocp commented:
Just looking through `bokeh/application/handlers` a bit to see when user code runs. I think code_runner.py `new_module` is safe to do pre-fork, since it makes a module object but does not execute the code in the module (?). The thing that must be post-fork is code_runner.py run().

One possibility to catch any future mistakes would be: at the start of code_runner.py run(), set some sort of "we have now run some user code" flag. 

Probably right before the `self._http.start(num_procs)` in the server, we'd then add an `assert no_user_code_has_been_run` based on this flag.

There could be a property like `safe_to_fork` on Application which chains to a property on ApplicationHandler which chains to a property on CodeRunner (for those handlers that have a CodeRunner), and server would assert `self._application.safe_to_fork` maybe ...

Anyway this is just a suggestion, @bryevdv is the arbiter of what he wants to do here. I may be overly paranoid, I'm just imagining someone in a year or two moving code around and accidentally making user handlers run pre-fork, which could cause a bunch of problems, so I'm trying to come up with a way to "prove" by runtime assertion that we don't do this.

Could also be an elaboration in a future PR of course, or just not worth doing.

birdsarah commented:
@boylea - are you available to keep working on this?

boylea commented:
Yes, I'll get to it tonight

boylea commented:
Made the recommended fixes.

I was going to manually explore the forking, but the dev environment in finicky, so I wasn't able to get very far

quasiben commented:
@boylea thanks for taking this over.  This is so awesome!

bryevdv commented:
@boylea Yes, thanks! You mentioned things being finicky, do you want to continue to pursue the "safe to fork" flag, with questions/help? If not that's OK, I'm ok with merging this as-is, as long as an issue is made for adding that sort of check separately.

bryevdv commented:
@boylea it looks like your other PR created a small merge conflict. Do you think you can resolved the conflict and push the update? Then we can make an issue to go back and add the "safe to fork" flag later, and get this PR merged.

boylea commented:
I added the assertion that user code should not run before forking multiple processes. I've tested that this flag works manually, but have *not* included an integration test to make sure the flag will fail if user code in run first.

Rebased to fix merge conflicts.

bryevdv commented:
All expected tests green, tested locally. Merging, thanks @boylea !
